server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java

```diff
@@ -0,0 +1,1051 @@
+        }
+        return backupScheduleDao.remove(schedule.getId());
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_VM_BACKUP_CREATE, eventDescription = "creating VM backup", async = true)
+    public boolean createBackup(final Long vmId) {
+        final VMInstanceVO vm = vmInstanceDao.findById(vmId);
+        if (vm == null) {
+            throw new CloudRuntimeException("Did not find VM by provided ID");
+        }
+        validateForZone(vm.getDataCenterId());
+        accountManager.checkAccess(CallContext.current().getCallingAccount(), null, true, vm);
+
+        if (vm.getBackupOfferingId() == null) {
+            throw new CloudRuntimeException("VM has not backup offering configured, cannot create backup before assigning it to a backup offering");
+        }
+
+        final BackupOffering offering = backupOfferingDao.findById(vm.getBackupOfferingId());
+        if (offering == null) {
+            throw new CloudRuntimeException("VM backup offering not found");
+        }
+
+        if (!offering.isUserDrivenBackupAllowed()) {
+            throw new CloudRuntimeException("The assigned backup offering does not allow ad-hoc user backup");
+        }
+
+        final BackupProvider backupProvider = getBackupProvider(offering.getProvider());
+        if (backupProvider != null && backupProvider.takeBackup(vm)) {
+            return true;
+        }
+        throw new CloudRuntimeException("Failed to create VM backup");
+    }
+
+    @Override
+    public Pair<List<Backup>, Integer> listBackups(final ListBackupsCmd cmd) {
+        final Long id = cmd.getId();
+        final Long vmId = cmd.getVmId();
+        final Long zoneId = cmd.getZoneId();
+        final Account caller = CallContext.current().getCallingAccount();
+        final String keyword = cmd.getKeyword();
+        List<Long> permittedAccounts = new ArrayList<Long>();
+
+        if (vmId != null) {
+            VMInstanceVO vm = vmInstanceDao.findByIdIncludingRemoved(vmId);
+            if (vm != null) {
+                accountManager.checkAccess(caller, null, true, vm);
+            }
+        }
+
+        final Ternary<Long, Boolean, Project.ListProjectResourcesCriteria> domainIdRecursiveListProject = new Ternary<Long, Boolean, Project.ListProjectResourcesCriteria>(cmd.getDomainId(),
+                cmd.isRecursive(), null);
+        accountManager.buildACLSearchParameters(caller, id, cmd.getAccountName(), cmd.getProjectId(), permittedAccounts, domainIdRecursiveListProject, cmd.listAll(), false);
+        final Long domainId = domainIdRecursiveListProject.first();
+        final Boolean isRecursive = domainIdRecursiveListProject.second();
+        final Project.ListProjectResourcesCriteria listProjectResourcesCriteria = domainIdRecursiveListProject.third();
+
+        final Filter searchFilter = new Filter(BackupVO.class, "id", true, cmd.getStartIndex(), cmd.getPageSizeVal());
+        SearchBuilder<BackupVO> sb = backupDao.createSearchBuilder();
```

Comment:
Same here below

Reply:
Will check.

Reply:
I looked into it, it's possible to move it to another DaoImpl, but then it would require moving the usage of `accountManager` and joins etc, that's moving the complexity to another class. I'm therefore keeping it all cohesive and tight in the backupmanagerimpl instead.

Reply:
Copy that